File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public IList<IList<int>> LevelOrder(TreeNode root) {
        IList<IList<int>> res = new List<IList<int>>();
        if (root == null) return res;
        Queue<TreeNode> q = new Queue<TreeNode>();
        q.Enqueue(root);
        
        while (q.Count > 0) {
            int qLen = q.Count;
            IList<int> level = new List<int>();
            
            for (int i = 0; i < qLen; i++) {
                TreeNode node = q.Dequeue();
                level.Add(node.val);
                if (node.left != null) q.Enqueue(node.left);
                if (node.right != null) q.Enqueue(node.right);
            }
            
            res.Add(level);
        }
        
        return res;
    }
}

File: hard/0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

// just BFS with a queue, but
// have to have inner loop to 
// empty items of level in queue
const levelOrder = function(root) {
    if (!root) return [];
    const levels = [];
    const q = [root];
    while (q.length) {
        const level = [];
        let levelLength = q.length;
        while (levelLength--) {
            const node = q.shift();
            level.push(node.val);
            node.left && q.push(node.left);
            node.right && q.push(node.right);
        }
        levels.push(level);
    }
    return levels;
};